File: src/lib/scoring.ts
import type { IndicatorResult } from "./indicators";
import type { ICTResult, TrendSignal } from "@/types";

export function scoreTimeframe(
  ind: IndicatorResult,
  ict: ICTResult
): { trend: TrendSignal; score: number } {
  let score = 0;

  // Classic indicators
  if (ind.rsi !== null) score += ind.rsi > 55 ? 1 : ind.rsi < 45 ? -1 : 0;
  if (ind.macd_hist !== null) score += ind.macd_hist > 0 ? 1 : -1;
  if (ind.macd !== null && ind.macd_signal !== null) score += ind.macd > ind.macd_signal ? 1 : -1;
  if (ind.close !== null && ind.ema_50 !== null) score += ind.close > ind.ema_50 ? 1 : -1;
  if (ind.close !== null && ind.ema_200 !== null) score += ind.close > ind.ema_200 ? 1 : -1;

  // ICT: price inside an Order Block
  if (ind.close !== null) {
    for (const ob of ict.orderBlocks) {
      if (ind.close >= ob.low && ind.close <= ob.high) {
        score += ob.type === "bullish" ? 1 : -1;
      }
    }
  }

  // ICT: liquidity sweep on last candle (stop hunt = likely reversal)
  if (ict.recentSSLSweep) score += 1; // swept lows → bullish reversal
  if (ict.recentBSLSweep) score -= 1; // swept highs → bearish reversal

  score = Math.max(-3, Math.min(3, score));
  const trend: TrendSignal = score >= 2 ? "BULLISH" : score <= -2 ? "BEARISH" : "NEUTRAL";
  return { trend, score };
}

export function aggregateTrends(trends: TrendSignal[]): {
  global_trend: TrendSignal;
  global_score: number;
  confluence: boolean;
} {
  const map: Record<TrendSignal, number> = { BULLISH: 1, NEUTRAL: 0, BEARISH: -1 };
  const total = trends.reduce((s, t) => s + map[t], 0);
  const confluence = new Set(trends).size === 1 && trends[0] !== "NEUTRAL";
  const global_trend: TrendSignal = total >= 2 ? "BULLISH" : total <= -2 ? "BEARISH" : "NEUTRAL";
  return { global_trend, global_score: total, confluence };
}

File: api/app/services/scoring.py
"""
Trend scoring engine.

Each timeframe produces a score from -3 to +3 based on 4 signals:
  +1 / -1  RSI > 55 / RSI < 45
  +1 / -1  MACD histogram > 0 / < 0
  +1 / -1  MACD line > signal line
  +1 / -1  Price > EMA-50
  +1 / -1  Price > EMA-200  (bonus weight for long-term trend)

Raw score ranges -5 to +5; we clamp to [-3, +3] for display.
"""
from app.models.schemas import IndicatorSnapshot, TrendSignal


def score_timeframe(snap: IndicatorSnapshot) -> tuple[TrendSignal, int]:
    score = 0

    if snap.rsi is not None:
        if snap.rsi > 55:
            score += 1
        elif snap.rsi < 45:
            score -= 1

    if snap.macd_hist is not None:
        score += 1 if snap.macd_hist > 0 else -1

    if snap.macd is not None and snap.macd_signal is not None:
        score += 1 if snap.macd > snap.macd_signal else -1

    if snap.close is not None and snap.ema_50 is not None:
        score += 1 if snap.close > snap.ema_50 else -1

    if snap.close is not None and snap.ema_200 is not None:
        score += 1 if snap.close > snap.ema_200 else -1

    score = max(-3, min(3, score))

    if score >= 2:
        trend: TrendSignal = "BULLISH"
    elif score <= -2:
        trend = "BEARISH"
    else:
        trend = "NEUTRAL"

    return trend, score


def aggregate_trends(tf_trends: list[TrendSignal]) -> tuple[TrendSignal, int, bool]:
    """
    Returns (global_trend, global_score, confluence).
    confluence = True only when all 3 timeframes share the same non-NEUTRAL trend.
    """
    score_map = {"BULLISH": 1, "NEUTRAL": 0, "BEARISH": -1}
    total = sum(score_map[t] for t in tf_trends)

    confluence = len(set(tf_trends)) == 1 and tf_trends[0] != "NEUTRAL"

    if total >= 2:
        global_trend: TrendSignal = "BULLISH"
    elif total <= -2:
        global_trend = "BEARISH"
    else:
        global_trend = "NEUTRAL"

    return global_trend, total, confluence
